### PR TITLE
Ability to authenticate with the schema registry via Cognito OAuth

### PR DESCRIPTION
**Pull request by joshuafleck**
- Ability to authenticate with the schema registry via Cognito OAuth
- Will fetch a bearer token to be passed with requests to the schema registry
- If the bearer token expires, it will fetch a new one
- It will retry up to 3 times to fetch a new bearer token before giving up

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -274,3 +274,17 @@ end
 
 # Messaging objects created with the same registry_url will now use the fake server.
 ```
+
+### Cognito (Oauth) Compatability
```
Comment (fbad) [inline README.md]: typo

File: lib/avro_turf/connection_wrapper_with_token_auth.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      client_key_pass: client_key_pass,
+      client_cert_data: client_cert_data,
+      client_key_data: client_key_data)
+  end
+
+  def with_connection
+    result = super
+    @refresh_token_retries_remaining = REFRESH_TOKEN_TRIES
+    result
+  rescue Excon::Error::Unauthorized
```
Comment (fbad) [inline lib/avro_turf/connection_wrapper_with_token_auth.rb:46]: Can't we call refresh before the request (it's just a if if refresh is not needed, right?)? I think we can reduce requests this way?

File: lib/avro_turf/messaging.rb [hunk 3]
```diff
@@ -89,6 +95,9 @@ def initialize(
           client_key_pass: client_key_pass,
           client_cert_data: client_cert_data,
           client_key_data: client_key_data,
+          oauth_url: oauth_url,
+          oauth_client_id: oauth_client_id,
+          oauth_client_secret: oauth_client_secret,
```
Comment (fbad) [inline lib/avro_turf/messaging.rb:100]: lib/avro_turf/messaging.rb#130 shall we add a config param to enable/disable automatic registration of schemas?

File: lib/avro_turf/connection_wrapper_with_token_auth.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                 client_key_pass: nil,
+                 client_cert_data: nil,
+                 client_key_data: nil,
+                 oauth_url: nil,
+                 oauth_client_id: nil,
+                 oauth_client_secret: nil)
+    @oauth_url = oauth_url
+    @oauth_client_id = oauth_client_id
+    @oauth_client_secret = oauth_client_secret
+    @semaphore = Mutex.new
```
Comment (mijoharas) [inline lib/avro_turf/connection_wrapper_with_token_auth.rb:23]: I'm only halfway through reviewing this, so I need to see how this class is used, but this mutex is created per instance. Is this class used as a singleton? or are we trying to protect cross thread activity of the same instance? will come back to this.

Comment (joshuafleck) [reply lib/avro_turf/connection_wrapper_with_token_auth.rb:23]: Yes, we maintain one instance of it when we use it, here: https://github.com/meetcleo/meetcleo/pull/23203/files#diff-46c130aaced44d9d3f3a294431826f310567402d64b4b388074ee7a0b9726828R20

Comment (joshuafleck) [reply lib/avro_turf/connection_wrapper_with_token_auth.rb:23]: It's so we can take advantage of the caching, here: https://github.com/meetcleo/avro_turf/blob/master/lib/avro_turf/messaging.rb#L79

**Review by fbad — APPROVED** [review]
LGTM

File: lib/avro_turf/connection_wrapper_with_token_auth.rb (hunk 4 above)
Comment (mijoharas) [reply lib/avro_turf/connection_wrapper_with_token_auth.rb:23]: perfect, I was actually just reviewing that now.

**Review by mijoharas — APPROVED** [review]

**Pull request merged**